compiler/src/iree/compiler/Dialect/Flow/Transforms/RaiseSpecialOps.cpp

```diff
@@ -351,6 +359,179 @@ static FailureOr<Operation *> tryRaiseToView(linalg::GenericOp linalgOp,
+//     %neg = arith.negf %extracted : f16
+//     %cmp = arith.cmpi eq, %in-1, %c1 : index
+//     %sel = arith.select %cmp, %neg, %extracted : f16
+//     linalg.yield %sel : f16
+//   } -> tensor<...x2x64xf16>
+//   %collapsed = tensor.collapse_shape %2 [[0], ..., [n-1, n]] : tensor<...x2x64xf16> into tensor<...x128xf16>
+//
+// clang-format on
+static std::optional<Value>
+matchCatNegateAndSlice(tensor::InsertSliceOp insertOp) {
+  /// First match against the desired op chain.
+  auto topHalf = insertOp.getSource().getDefiningOp<tensor::ExtractSliceOp>();
+  if (!topHalf || !topHalf->hasOneUse()) {
+    return std::nullopt;
+  }
+  auto dest = insertOp.getDest().getDefiningOp<tensor::InsertSliceOp>();
+  if (!dest || !dest->hasOneUse()) {
+    return std::nullopt;
+  }
+  auto outEmpty = dest.getDest().getDefiningOp<tensor::EmptyOp>();
+  if (!outEmpty) {
+    return std::nullopt;
+  }
+
+  /// TODO: This could be extended to other unary (or in general, elementwise
+  /// operations) if the need arises.
+  auto negate = dest.getSource().getDefiningOp<linalg::GenericOp>();
+  if (!negate || !isUnaryNegate(negate)) {
+    return std::nullopt;
+  }
+
+  auto bottomHalf =
+      negate.getDpsInputs()[0].getDefiningOp<tensor::ExtractSliceOp>();
+  if (!bottomHalf || !bottomHalf->hasOneUse()) {
+    return std::nullopt;
+  }
+  Value source = topHalf.getSource();
+  if (source != bottomHalf.getSource()) {
+    return std::nullopt;
+  }
+
+  /// Require that the overall operation isn't changing the tensor shape.
+  if (source.getType() != outEmpty.getType()) {
+    return std::nullopt;
+  }
+
+  auto sourceType = source.getType().dyn_cast<RankedTensorType>();
+  if (!sourceType) {
+    return std::nullopt;
+  }
+
+  /// Here we are verifying that the slice sequence inserts and extracts in the
+  /// same order. In other words, the offsets, sizes, and strides for the nth
+  /// extract slice match the nth insert slice.
+  /// TODO: We could relax this condition if we see cases where the elementwise
+  /// unary operation happens on a different slices of the tensor, but for now
+  /// don't engineer for cases that don't exist.
+  auto isSame = [](OpFoldResult a, OpFoldResult b) { return a == b; };
+  if (!insertOp.isSameAs(bottomHalf, isSame) ||
+      !dest.isSameAs(topHalf, isSame)) {
+    return std::nullopt;
+  }
+
+  /// All slices need unit stride and the initial slice should have a zero
+  /// offset.
+  if (!insertOp.hasUnitStride() || !dest.hasUnitStride() ||
+      !dest.hasZeroOffset()) {
+    return std::nullopt;
+  }
+
+  /// Verify all slice sizes are the same.
+  for (auto [bottom, top] :
+       llvm::zip_equal(insertOp.getMixedSizes(), dest.getMixedSizes())) {
+    if (bottom != top) {
+      return std::nullopt;
+    }
+  }
+
+  /// Note that we already verified all offsets, sizes, and strides are the same
+  /// for the final insert and bottom extract, so this checks both.
```

Comment:
Comment is a bit unclear. Also can be written as

```
if (llvm::any_of(insertOp.getMixedOffset(), [](OpFoldResult ofr) { return !isConstantIntValue(ofr, 0);})) {
  return std::nullopt;
}
```

Reply:
The loop only checks up to the second-to-last offset so I just changed to using `isConstantIntValue` and adjusted the comment.